lib/Core/Executor.cpp

```diff
@@ -1483,33 +1487,52 @@ void Executor::executeCall(ExecutionState &state,
         return;
       }
 
+      // compute widths of arguments: if the argument is byval, we use the
+      // pointee width
+      uint64_t argWidths[callingArgs];
+      CallSite cs(i);
+      for (unsigned k = 0; k < callingArgs; k++) {
+        if (cs.isByValArgument(k)) {
+#if LLVM_VERSION_CODE >= LLVM_VERSION(9, 0)
+          Type *t = cs.getParamByValType(k);
+#else
+          auto arg = cs.getArgument(k);
+          Type *t = arg->getType();
+          assert(t->isPointerTy());
+          t = t->getPointerElementType();
+#endif
+          argWidths[k] = kmodule->targetData->getTypeSizeInBits(t);
+        } else {
+          argWidths[k] = arguments[k]->getWidth();
+        }
+      }
+
       StackFrame &sf = state.stack.back();
       unsigned size = 0;
       bool requires16ByteAlignment = false;
-      for (unsigned i = funcArgs; i < callingArgs; i++) {
+      for (unsigned k = funcArgs; k < callingArgs; k++) {
         // FIXME: This is really specific to the architecture, not the pointer
         // size. This happens to work for x86-32 and x86-64, however.
         if (WordSize == Expr::Int32) {
-          size += Expr::getMinBytesForWidth(arguments[i]->getWidth());
+          size += Expr::getMinBytesForWidth(argWidths[k]);
         } else {
-          Expr::Width argWidth = arguments[i]->getWidth();
           // AMD64-ABI 3.5.7p5: Step 7. Align l->overflow_arg_area upwards to a
           // 16 byte boundary if alignment needed by type exceeds 8 byte
           // boundary.
           //
           // Alignment requirements for scalar types is the same as their size
-          if (argWidth > Expr::Int64) {
+          if (argWidths[k] > Expr::Int64 && !cs.isByValArgument(k)) {
```

Comment:
I'm not sure if the `!cs.isByValArgument()` is correct - have a look at the email I sent you.

Reply:
Sure, we can chat over email about this; let me know if you're understanding of the spec is different.